akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -699,18 +790,63 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
    *
    * This object can be cached and reused within the same [[GraphStageLogic]].
    */
-  final def getAsyncCallback[T](handler: T ⇒ Unit): AsyncCallback[T] = {
+  final protected def getAsyncCallback[T](handler: T ⇒ Unit): AsyncCallback[T] = {
     new AsyncCallback[T] {
       override def invoke(event: T): Unit =
         interpreter.onAsyncInput(GraphStageLogic.this, event, handler.asInstanceOf[Any ⇒ Unit])
     }
   }
 
+  // TODO rethink if it's worth it; move the Updater to an object; etc
+  // via atomics since users may do something weird - try to Future { init } or something...
+  private final val _setStageActorRef = AtomicReferenceFieldUpdater.newUpdater(classOf[GraphStageLogic], classOf[StageActorRef], "_stageActorRefDoNotCallMeDirectly")
```

Comment:
we should simply disallow calling the getActorRef method from outside the stage.